### DISCUSSION TITLE
Does't  it possible to run tests in parallel against multiple environments in the same time?

### DISCUSSION THREAD

**Discussion by qalinn** [cypress-io/cypress#15074] [Questions and Help]
Hello! I have a question regarding parallelization. I want to run my tests in parallel, I noticed cypress can do something to run the tests in parallel, but it solves just half of the problem because cypress is running in parallel against only one environment. I want to do something like this: I have 3 tests(testA, TestB, TestC) and two environments(Env1 and Env2). Is it possible to configure cypress to execute the tests in parallel against these two environments at the same time? In java, I did for TestNG and Junit.
Thank you!

**Comment by bahmutov** [comment]
Yes, it is possible. Read this blog post that runs the same tests in parallel for different browser timezones. Using different environments could be done in the same way https://glebbahmutov.com/blog/testing-timezones/ or even simpler if you have config files. Something like this:

```
- npx cypress run --config-file cypress-env1.json --record --parallel --group Env1
- npx cypress run --config-file cypress-env2.json --record --parallel --group Env2
```

**Reply by qalinn** [reply]
Hi @bahmutov, 

Thank you very much for the explanations and for the link.
I have checked the resources but my problem is still there.  Let me try to explain again what I want to achieve.
So, I have a few tests and these  are going on web page and check if a button is there. 
Test 1 is going on http://testA.com and is checking if Button1 is there
Test 2 is going on http://testB.com and is checking if Button2 is there
Test 3 is going on http://testC.com and is checking if Button3 is there

That means I have 3 test environments and I want to run these tests in parallel against these 3 environments.
I don't want to make by hand groups and set them manually. Should be able to pass the name of enivronment-urls(testA,testB, testC) and cypress should take this list and start to run the tests in parallel for all these environments and for all tests(if you have 10 tests, cypress will take the first 3 tests run against the test environments(that can be translated into an http url) when a test is finished, cypress runner should take the next test and run against the available environment). 

Thank you!

**Reply by bahmutov** [reply]
My answer still stands. You have to make the groups, especially if the tests are going to use different config files. But all tests in a group will be running in parallel, which is great. Bonus tip: you can use https://github.com/bahmutov/cypress-select-tests plugin to skip some tests depending on the environment or run only on selected environment.

**Comment by qalinn** [comment]
Hi @bahmutov, 

Sorry because I insist on this, but it can be a nice feature. Imagine like this you have the same version of an app. This app is deployed on 3 machines. Before/after every test you want to reset the db or something else to leave the environment in the same state for all your tests. If you are doing it in parallel with cypress that is possible to broke other tests. if it will be possible to set at runtime the base_url for your tests you can avoid this.
Please find attached a small picture how should happen for 3 targets.
![image](https://user-images.githubusercontent.com/6930622/108056478-99354880-7051-11eb-967f-a221ee593578.png)

**Reply by bahmutov** [reply]
I still don't understand what you need. You can vary the baseUrl, you can even change it a run-time. But if you want to run all specs (or some of them) against the baseUrl, then run the tests against another baseUlr, etc - then what I have suggested is good solution.

**Reply by qalinn** [reply]
Hello @bahmutov, 

I am trying to do  the following: 
1) I have a set of specs/tests (20 specs).
2) I will deploy my app on 4 machines(the same version of the app). Just to help me to run my tests in parallel. These machines can be different(for example spot instances from amazon, different IPs, different machines all the time)
3) I will collect the IP of every machine.= IP list 
4) With the IP list  I am going back to cypress and I want to pass this list of ip to cypress to let knows that: I have 4 IPs and I want to run all my specs/tests against these IPs.
5) Cypress should be able to take the list  and should prepare the first 4 specs with base_url according to the 4 IPs received at the beginning(this list of IPs all the time will be a new one)
6) Cypress runner will start to run the first 4 specs in parallel against 4 different machines.
7) When the first spec will finish, that means the machine is available to take another spec to run. Cypress should know that and will take the next spec and run on that free machine.
8) All the tests will run in the same way till the end. When a machine is available the runner will take a new spec and will configure with the base_url with the right IP and will start to run.

I hope now I was a little bit clear.

If you want we can set a meeting to explain this to you in more detail about this feature.
My email address is: qalinn@yahoo.com
